packages/chat-component/src/main.ts

```diff
@@ -377,96 +368,345 @@ export class ChatComponent extends LitElement {
+    const { text: lastChatMessageTextEntry } = this.chatMessages.at(-1) as ChatMessage;
+
+    if (!lastChatMessageTextEntry[textBlockIndex]) {
+      lastChatMessageTextEntry[textBlockIndex] = {
+        value: '',
+        followingSteps: [],
+      };
+    }
+    lastChatMessageTextEntry[textBlockIndex].value += chunkValue;
+  }
+
+  // update the following steps or followup questions entry
+  updateFollowingStepOrFollowupQuestionEntry({
+    chunkValue,
+    textBlockIndex,
+    stepIndex,
+    isFollowupQuestion,
+  }: {
+    chunkValue: string;
+    textBlockIndex: number;
+    stepIndex: number;
+    isFollowupQuestion: boolean;
+  }) {
+    // following steps and followup questions are treated the same way. They are just stored in different arrays
+    const { followupQuestions, text: lastChatMessageTextEntry } = this.chatMessages.at(-1) as ChatMessage;
+    if (isFollowupQuestion && followupQuestions) {
+      followupQuestions[stepIndex] = (followupQuestions[stepIndex] || '') + chunkValue;
+      return;
+    }
+
+    if (lastChatMessageTextEntry && lastChatMessageTextEntry[textBlockIndex]) {
+      const { followingSteps } = lastChatMessageTextEntry[textBlockIndex];
+      if (followingSteps) {
+        followingSteps[stepIndex] = (followingSteps[stepIndex] || '') + chunkValue;
+      }
+    }
+  }
+
+  // Add a message to the chat, when the user or the API sends a message
+  async addMessage(message: string, isUserMessage: boolean) {
+    const citations: Citation[] = [];
+    const followingSteps: string[] = [];
+    const followupQuestions: string[] = [];
+    // Get the timestamp for the message
+    const timestamp = this.getTimestamp();
+    const updateChatWithMessageOrChunk = async (part: string, isChunk: boolean) => {
+      if (isChunk) {
+        await this.consumeStreamedMessage({
+          timestamp,
+          isUserMessage,
+        });
+        return true;
+      }
+
+      this.chatMessages = [
+        ...this.chatMessages,
+        {
+          text: [
+            {
+              value: part,
+              followingSteps,
+            },
+          ],
+          followupQuestions,
+          citations: [...new Set(citations)],
+          timestamp: timestamp,
+          isUserMessage,
+        },
+      ];
+      return true;
+    };
+
+    // Check if message is a bot message to process citations and follow-up questions
+    if (isUserMessage) {
+      updateChatWithMessageOrChunk(message, false);
+    } else {
+      if (this.isStreaming) {
+        await updateChatWithMessageOrChunk(message /* undefined */, true);
+        this.requestUpdate('chatMessages');
+      } else {
+        // non-streamed response
+        const processedText = processText(message, [citations, followingSteps, followupQuestions]);
+        message = processedText.replacedText;
+        // Push all lists coming from processText to the corresponding arrays
+        citations.push(...(processedText.arrays[0] as unknown as Citation[]));
```

Comment:
Would it be possible to properly type the processText return to avoid these casts?

Reply:
we will probably remove this part of the code since now we are mainly processing streams.